core/src/main/java/org/apache/calcite/prepare/CalciteCatalogReader.java

```diff
@@ -156,9 +171,17 @@ private Procedure getProcedure(List<String> names) {
       }
       final String name = Util.last(names);
       boolean caseSensitive = nameMatcher.isCaseSensitive();
-      Procedure procedure = schema.getProcedure(name, caseSensitive);
-      if (procedure != null) {
-        return procedure;
+      Function function = null;
+      switch (category) {
+      case USER_DEFINED_PROCEDURE:
+        function = schema.getProcedure(name, caseSensitive);
+        break;
+      case USER_DEFINED_MACRO:
+        function = schema.getMacro(name, caseSensitive);
+        break;
+      }
+      if (function != null) {
```

Comment:
nit: Can we remove this and just `return function` unconditionally at the end of this function?

Reply:
I'm not 100% sure on this one. All tests pass if I return function at the end but I don't know if that will work for all cases. Could it happen that function is not null and then later becomes null via more iterations of the loop?

Reply:
Ah I see what you mean. In that case, let's just leave it as is.